### PR TITLE
Fix: stacked bar legend toggle(fixes #283)

### LINKED ISSUE #283: Stacked bar legend toggle
On the carbon site, the stacked bar legend toggle is not working as anticipated. When a legend item is toggled off, the associated portion of the bars are not being removed properly from the graph. i.e. when you toggle off the blue, it moves the purple portion of the bar down and does not remove the blue portion.  
Screen Shot 2020-07-31 at 10 40 50 AM

Also, when the first legend item is toggled off, nothing happens. 
Screen Shot 2020-07-31 at 10 40 15 AM
**Proposed Design Fix**
When blue is toggled off - blue is removed from the stacked bar and all other portions adjust accordingly. 
Screen Shot 2020-07-31 at 11 00 22 AM

### PR DESCRIPTION
**Pull request by AshishMotanamGurunadham**
Fixes: #283 

**What is the purpose of this pull request? (Check any that's applicable)**

-   [ ] Feature
-   [x] Bug fix
-   [ ] Dependency updates
-   [ ] Documentation updates
-   [ ] Build related changes
-   [ ] Changes an existing functionality
-   [ ] Other, please explain:

**Does this introduce a breaking change?**

-   [ ] Yes
-   [x] No

**What changes did you make? (Give an overview)**
- Fixed ongoing legend toggle issue in stacked bar graph and grouped bar graph.
- Fixed Grouped goal line bar graph page loading issue.
- Fixed regions issue in grouped goal line bar graph.
 
**Add any screenshots**

- Stacked bar legend toggle, when Data label 1 is clicked: 
   https://engineering.cerner.com/carbon-graphs/#/bar/stacked-bars

**Before**
image

**After**
image

- Grouped bar legend toggle, when Data label 2 is clicked: 
   https://engineering.cerner.com/carbon-graphs/#/bar/grouped-bars

**Before**
image

**After**
image

- Grouped goal line bar graph: 
   https://engineering.cerner.com/carbon-graphs/#/regions/bar/grouped-goal-lines

**Before**
image

**After**
image

- Regions in Grouped goal line bar graph: 
   https://engineering.cerner.com/carbon-graphs/#/regions/bar/grouped-goal-lines

**Before**
image

**After**
image

### CODE REVIEW

**Review by jeremyfuksa — APPROVED** [review]

File: src/test/unit/controls/Bar/StackedBar-spec.js [hunk 1]
```diff
@@ -132,6 +133,52 @@ describe("Stacked Bar", () => {
             );
             expect(barsContainer.length).toEqual(0);
         });
+        it("Hide Bar content on single click", (done) => {
```
Comment (sdadn) [inline src/test/unit/controls/Bar/StackedBar-spec.js:136]: nit: hides

File: src/test/unit/controls/Bar/GroupedBar-spec.js [hunk 2]
```diff
@@ -129,6 +130,52 @@ describe("Grouped Bar", () => {
             );
             expect(barsContainer.length).toEqual(0);
         });
+        it("Hide Bar content on single click", (done) => {
```
Comment (sdadn) [inline src/test/unit/controls/Bar/GroupedBar-spec.js:133]: nit: _hides_

File: src/test/unit/controls/Bar/StackedBar-spec.js [hunk 3]
```diff
@@ -132,6 +133,52 @@ describe("Stacked Bar", () => {
...
+            });
+            triggerEvent(legendItem[1], "click", () => {
+                barsContainer;
+                expect(barsContainer[0].getAttribute("aria-hidden")).toBe(
+                    "true"
+                );
+                done();
+            });
+        });
+        it("Dispaly Bar content on double click", (done) => {
```
Comment (sdadn) [inline src/test/unit/controls/Bar/StackedBar-spec.js:162]: nit: _displays_

**Review by sdadn — CHANGES_REQUESTED** [review]

File: src/test/unit/controls/Bar/StackedBar-spec.js (hunk 1 above)
Comment (AshishMotanamGurunadham) [reply src/test/unit/controls/Bar/StackedBar-spec.js]: [Fixed](https://github.com/cerner/carbon-graphs/pull/300/commits/6678d3c6f4bb3056543ae2945ffa3f922a93648e#diff-70b14f0c2ecf38ac198e67378c9b8b42R136)

File: src/test/unit/controls/Bar/StackedBar-spec.js (hunk 3 above)
Comment (AshishMotanamGurunadham) [reply src/test/unit/controls/Bar/StackedBar-spec.js]: [Fixed](https://github.com/cerner/carbon-graphs/pull/300/commits/6678d3c6f4bb3056543ae2945ffa3f922a93648e#diff-70b14f0c2ecf38ac198e67378c9b8b42R162)

File: src/test/unit/controls/Bar/GroupedBar-spec.js (hunk 2 above)
Comment (AshishMotanamGurunadham) [reply src/test/unit/controls/Bar/GroupedBar-spec.js]: [Fixed](https://github.com/cerner/carbon-graphs/pull/300/commits/6678d3c6f4bb3056543ae2945ffa3f922a93648e#diff-c4ce2bdb663a50ba09b690a4c58dacdbR133)

File: dev/examples/controls/bar.js [hunk 4]
```diff
@@ -71,6 +71,32 @@ const regions2 = [
         end: 30
     }
 ];
+const regions3 = [
```
Comment (mraplumb) [inline dev/examples/controls/bar.js:74]: nit: what is the difference between `regions`, `regions2` and `regions3`.  can we use a better variable name here?

File: src/main/js/controls/Bar/helpers/legendHelpers.js [hunk 5]
```diff
@@ -47,16 +47,55 @@ const clickHandler = (graphContext, control, config, canvasSVG) => (
...
+                return !(d3.select(this).attr("aria-hidden") === "true");
+            });
+        setSelectionIndicatorAttributes(
+            canvasSVG.selectAll(`.${styles.taskBarSelection}`),
+            false
+        );
+        setSelectionIndicatorAttributes(
+            canvasSVG.selectAll(`.${styles.taskBarSelection}`),
+            false
+        );
```
Comment (mraplumb) [inline src/main/js/controls/Bar/helpers/legendHelpers.js:80]: Is it intentional that we have this code duplicated?

File: src/main/js/controls/Bar/helpers/legendHelpers.js [hunk 6]
```diff
@@ -47,16 +47,55 @@ const clickHandler = (graphContext, control, config, canvasSVG) => (
...
+        window.requestAnimationFrame(onAnimationHandler(graphContext));
+    } else {
```
Comment (mraplumb) [inline src/main/js/controls/Bar/helpers/legendHelpers.js:82]: Doesn't this `else` block cover the same condition as the initial `if(config.shownTargets.length === 0)`?  Can these two blocks be consolidated?

**Review by mraplumb — CHANGES_REQUESTED** [review]

File: src/test/unit/controls/Bar/BarRegions-spec.js [hunk 7]
```diff
@@ -590,6 +590,46 @@ describe("Bar - Region", () => {
                     });
                 });
             });
+            it("Hide region on single legend click", (done) => {
```
Comment (sdadn) [inline src/test/unit/controls/Bar/BarRegions-spec.js:593]: nit: hides

File: src/main/js/controls/Bar/helpers/legendHelpers.js [hunk 8]
```diff
@@ -47,16 +47,55 @@ const clickHandler = (graphContext, control, config, canvasSVG) => (
             shownTargets.push(item.key);
         }
     };
+    if (config.shownTargets.length === 0) {
+        graphContext.contentKeys.forEach((i) => {
```
Comment (sdadn) [inline src/main/js/controls/Bar/helpers/legendHelpers.js:51]: Would something other than _i_ be a better variable name? e.g. _key_

**Review by sdadn — CHANGES_REQUESTED** [review]

File: src/main/js/controls/Bar/helpers/legendHelpers.js (hunk 6 above)
Comment (AshishMotanamGurunadham) [reply src/main/js/controls/Bar/helpers/legendHelpers.js:82]: Yes but that if statement (if(config.shownTargets.length === 0)) needs to get executed before the on click execution gets started. That is the reason why written that separately.

File: dev/examples/controls/bar.js (hunk 4 above)
Comment (AshishMotanamGurunadham) [reply dev/examples/controls/bar.js]: @mraplumb  Those are different set of regions that we use in different scenarios. Removed region2 and region3 and added everything in single array(regions). you can see changes in [here](https://github.com/cerner/carbon-graphs/pull/300/commits/595865c1ef97c1cd837971b727d58b1e7d89c5c0#diff-5768493ee06baddc838b70d473847bb6R22).

File: src/main/js/controls/Bar/helpers/legendHelpers.js (hunk 5 above)
Comment (AshishMotanamGurunadham) [reply src/main/js/controls/Bar/helpers/legendHelpers.js]: that is not intentional Andy and i removed duplicate code [here](https://github.com/cerner/carbon-graphs/pull/300/commits/595865c1ef97c1cd837971b727d58b1e7d89c5c0#diff-1ecc848586009dc5a2a7976d5f75492eL77-L80).

File: src/main/js/controls/Bar/helpers/legendHelpers.js (hunk 8 above)
Comment (AshishMotanamGurunadham) [reply src/main/js/controls/Bar/helpers/legendHelpers.js]: [Fixed](https://github.com/cerner/carbon-graphs/pull/300/commits/595865c1ef97c1cd837971b727d58b1e7d89c5c0#diff-1ecc848586009dc5a2a7976d5f75492eR51-R53)

File: src/test/unit/controls/Bar/BarRegions-spec.js (hunk 7 above)
Comment (AshishMotanamGurunadham) [reply src/test/unit/controls/Bar/BarRegions-spec.js]: [Fixed](https://github.com/cerner/carbon-graphs/pull/300/commits/595865c1ef97c1cd837971b727d58b1e7d89c5c0#diff-1f108def83b42bc073df46dd9517a071R593)

**Review by mraplumb — APPROVED** [review]

File: src/main/js/controls/Bar/helpers/legendHelpers.js (hunk 6 above)
Comment (Veera81) [reply src/main/js/controls/Bar/helpers/legendHelpers.js:78]: All the above code are present in the `clickHandler` function

- How does if statement (if(config.shownTargets.length === 0)) executed before we click on the onclick?

- When will the else statement execute?

File: src/main/js/controls/Bar/helpers/legendHelpers.js [hunk 9]
```diff
@@ -47,16 +47,51 @@ const clickHandler = (graphContext, control, config, canvasSVG) => (
             shownTargets.push(item.key);
         }
     };
+    if (config.shownTargets.length === 0) {
```
Comment (Veera81) [inline src/main/js/controls/Bar/helpers/legendHelpers.js:50]: Can you add a comment on why we are checking here?

**Review by sdadn — APPROVED** [review]

**Review by Veera81 — APPROVED** [review]

**Pull request merged**